src/common/redis_module/ray_redis_module.c

```diff
@@ -272,6 +273,186 @@ int ResultTableLookup_RedisCommand(RedisModuleCtx *ctx,
+                        node_id, NULL);
+  } else {
+    RedisModule_HashSet(key, REDISMODULE_HASH_CFIELDS, "state", state, "node",
+                        node_id, "spec", spec, NULL);
+  }
+
+  /* Build the PUBLISH message for task table subscribers. This is a string in
+   * the format: "<task ID> <state> <node ID> <task specification>". */
+  const char *publish_field;
+  publish_field = RedisModule_StringPtrLen(state, &length);
+  if (length != 1) {
+    return RedisModule_ReplyWithError(ctx, "Invalid scheduling state width (must have width 1)");
+  }
+  RedisModule_StringAppendBuffer(ctx, task_id, " ", 1);
+  RedisModule_StringAppendBuffer(ctx, task_id, publish_field, length);
+  publish_field = RedisModule_StringPtrLen(node_id, &length);
+  RedisModule_StringAppendBuffer(ctx, task_id, " ", 1);
+  RedisModule_StringAppendBuffer(ctx, task_id, publish_field, length);
+  if (spec != NULL) {
+    publish_field = RedisModule_StringPtrLen(spec, &length);
+    RedisModule_StringAppendBuffer(ctx, task_id, " ", 1);
+    RedisModule_StringAppendBuffer(ctx, task_id, publish_field, length);
+  }
+
+  RedisModuleCallReply *reply =
+      RedisModule_Call(ctx, "PUBLISH", "ss", prefixed_keyname, task_id);
+  if (reply == NULL) {
+    return RedisModule_ReplyWithError(ctx, "PUBLISH unsuccessful");
+  }
+
+  RedisModule_CloseKey(key);
+  RedisModule_ReplyWithSimpleString(ctx, "ok");
+
+  return REDISMODULE_OK;
+}
+
+/**
+ * Add a new entry to the task table. This will overwrite any existing entry
+ * with the same task ID.
+ *
+ * This is called from a client with the command:
+ *
+ *     RAY.task_table_add <task ID> <state> <node ID> <task spec>
+ *
+ * @param task_id A string that is the ID of the task.
+ * @param state A string that is the current scheduling state (a
+ *        scheduling_state enum instance).
+ * @param node_id A string that is the ID of the associated node, if any.
+ * @param task_spec A string that is the specification of the task, which can
+ *        be cast to a `task_spec`.
+ * @return OK if the operation was successful.
+ */
+int TaskTableAddTask_RedisCommand(RedisModuleCtx *ctx,
+                                  RedisModuleString **argv,
+                                  int argc) {
+  if (argc != 5) {
+    return RedisModule_WrongArity(ctx);
+  }
+
+  return TaskTableWrite_RedisCommand(ctx, argv[1], argv[2], argv[3], argv[4]);
+}
+
+/**
+ * Update an entry in the task table. This does not update the task
+ * specification in the table.
+ *
+ * This is called from a client with the command:
+ *
+ *     RAY.task_table_update_task <task ID> <state> <node ID>
+ *
+ * @param task_string A string which is the task instance.
+ * @return OK if the operation was successful.
+ */
+int TaskTableUpdate_RedisCommand(RedisModuleCtx *ctx,
+                                 RedisModuleString **argv,
+                                 int argc) {
+  if (argc != 4) {
```

Comment:
We can't use getopt here, this is an internal redis convention to pass strings (these are not char *). It just uses the same convention as in main() but is really a different mechanism that wouldn't be compatible with getopt.